Multiple ingress classes

erwbgy opened the pull request:
Closes https://github.com/projectcontour/contour/issues/4097 Allow the contour --ingress-class-name value to be a comma-separated list of classes to match against.

stevesloka reviewed (dismissed):
- Just one small nit on the release note
- Maybe we should add some validation of the ingressClassName field now to make sure it is a comma separated list? I guess worst case it validates against the wrong string, but might be difficult to figure that out? 
- Would be neat to have some tests to show say two ingress objects, two httpproxy that match multiple classes.

youngnick commented:
@erwbgy, have you had a chance to check out the e2e tests?

erwbgy commented:
> @erwbgy, have you had a chance to check out the e2e tests?

Sorry I haven't and will be unlikely to be able to do so before Friday at the earliest.  If anyone else can see what the issues are and fix, please do!

skriss commented on test/e2e/httpproxy/multiple_ingress_classes_test.go:
```diff
@@ -0,0 +1,121 @@
+
+		f.Fixtures.Echo.Deploy(namespace, "ingress-conformance-echo")
+
+		for _, class := range []string{"", "contour", "team1"} {
```
An empty ingress class for the HTTPProxy wouldn't be valid here since Contour has been explicitly configured to process specific ingress classes.

erwbgy replied:
Thanks @skriss ! Such a simple fix but I couldn't see it.

skriss commented on test/e2e/httpproxy/multiple_ingress_classes_test.go:
```diff
@@ -0,0 +1,121 @@
+
+		f.Fixtures.Echo.Deploy(namespace, "ingress-conformance-echo")
+
+		for _, class := range []string{"", "contour", "team1"} {
```
ditto re: empty ingress class

erwbgy replied:
Fixed, thanks.

skriss commented on test/e2e/ingress/ingress_class_test.go:
```diff
@@ -0,0 +1,76 @@
+				Namespace: namespace,
+				Name:      name,
+			},
+			Spec: networkingv1.IngressSpec{
```
Need to specify an ingress class name here for the same reason as with HTTPProxy.

erwbgy replied:
Fixed, thanks.

erwbgy commented on changelogs/unreleased/4109-erwbgy-small.md:
```diff
@@ -0,0 +1 @@
+Allow the contour --ingress-class-name value to be a comma-separated list of classes to match against.
```
Fixed e2e tests with help from @skriss

youngnick commented:
This is looking pretty good to me, we'll need one of the other maintainers to take a look.

skriss commented on apis/projectcontour/v1alpha1/contourconfig.go:
```diff
@@ -138,9 +138,9 @@ type TLS struct {
+	// Ingress Class Names Contour should use.
 	// +optional
-	ClassName *string `json:"className,omitempty"`
+	ClassName []string `json:"className,omitempty"`
```
May as well update the names while the API is alpha:
```suggestion
	ClassNames []string `json:"classNames,omitempty"`
```

skriss commented on changelogs/unreleased/4109-erwbgy-small.md:
```diff
@@ -0,0 +1 @@
+Allow the contour --ingress-class-name value to be a comma-separated list of classes to match against.  Contour will process Ingress and HTTPProxy objects with any of the specified ingress classes.
```
Even though the ContourConfiguration CRD is alpha, I think it's worth calling out in the release note that the name (if you accept my previous comment) and type of the ingress class field are changing in a non-backwards-compatible way.

erwbgy replied:
Updated release note.

skriss commented on internal/ingressclass/ingressclass.go:
```diff
@@ -54,6 +56,11 @@ func matches(objIngressClass, contourIngressClass string) bool {
-	// Otherwise, the object's ingress class must match Contour's.
-	return objIngressClass == contourIngressClass
+	// Otherwise, the object's ingress class must match one of Contour's.
+	for _, contourIngressClassEntry := range strings.Split(contourIngressClass, ",") {
```
let's add a test case to https://github.com/projectcontour/contour/blob/main/internal/ingressclass/ingressclass_test.go#L26 and https://github.com/projectcontour/contour/blob/main/internal/ingressclass/ingressclass_test.go#L111 for a comma-separated `contourIngressClass`.

skriss commented on internal/dag/cache.go:
```diff
@@ -94,8 +95,10 @@ func (kc *KubernetesCache) matchesIngressClass(obj *networking_v1.IngressClass)
 	// configured.
-	if obj.Name == kc.IngressClassName {
-		return true
+	for _, contourIngressClassEntry := range strings.Split(kc.IngressClassName, ",") {
```
I don't think we're actually using the DAG cache's `ingressclass` field (https://github.com/projectcontour/contour/blob/81cbcabd93b5051e6c46ae0f122089a4fb8af926/internal/dag/cache.go#L54), which this function `matchesIngressClass` provides a filter for. Since Contour can now process >1 ingress class, I'd rather drop the field entirely, and this function, if it's truly not being used; otherwise, it would make sense that it'd need to be converted to a map or slice to support multiple ingress classes.

erwbgy replied:
I don't feel confident making DAG changes beyond simple ones like this.  Could this cleanup be deferred to a different MR?

skriss replied:
Yeah, we can do that, I'll create an issue for this.

skriss reviewed (changes requested):
Thanks for fixing those tests @erwbgy. Had a couple of additional comments

skriss commented:
@erwbgy just checking in to see if you'll be able to update this PR to resolve the conflicts and address the few comments, would be great to get this in before the next release!

erwbgy commented:
/remove lifecycle/stale

erwbgy commented:
@skriss @stevesloka I think I have addressed all of your comments.

Looks like the flake mentioned in https://github.com/projectcontour/contour/issues/4305 caused the e2e failure.

skriss commented on test/e2e/httpproxy/httpproxy_test.go:
```diff
@@ -428,3 +456,20 @@ descriptors:
 	})
 })
+
+// httpProxyValid returns true if the proxy has a .status.currentStatus
+// of "valid".
+func httpProxyValid(proxy *contourv1.HTTPProxy) bool {
+
+	if proxy == nil {
+		return false
+	}
+
+	if len(proxy.Status.Conditions) == 0 {
+		return false
+	}
+
+	cond := proxy.Status.GetConditionFor("Valid")
+	return cond.Status == "True"
+
+}
```
this function has moved into test/e2e/framework.go (`HTTPProxyValid`), so let's drop this copy and use that one instead.

skriss replied:
I can do this along with #4327 as a follow-up since it's just cleanup.

skriss reviewed (commented):
Thanks for the updates @erwbgy, just one more small thing from me.

skriss commented:
> Looks like the flake mentioned in https://github.com/projectcontour/contour/issues/4305 caused the e2e failure.

Yep, looks like it, passed on a re-run.

skriss reviewed (approved):
LGTM, thanks @erwbgy!
